src/pulselistener/pulselistener/mercurial.py

```diff
@@ -80,20 +84,43 @@ def cleanup(self):
             assert isinstance(diff, dict)
             assert 'phid' in diff
 
+            failure = None
+            start = time.time()
             try:
-                await self.handle_diff(diff)
+                await self.handle_diff(build_target_phid, diff)
             except hglib.error.CommandError as e:
                 logger.warn('Mercurial error on diff', error=e.err, args=e.args, phid=diff['phid'])
 
-                # Remove uncommited changes
-                self.repo.revert(self.repo_dir.encode('utf-8'), all=True)
+                # Report mercurial failure as a Unit Test issue
+                failure = UnitResult(
+                    namespace='code-review',
+                    name='mercurial',
+                    result=UnitResultState.Fail,
+                    details='WARNING: The code review bot failed to apply your patch.\n\n```{}```'.format(e.err),
+                    format='remarkup',
+                    duration=time.time() - start,
+                )
 
             except Exception as e:
                 logger.warn('Failed to process diff', error=e, phid=diff['phid'])
 
+                # Report generic failure as a Unit Test issue
+                failure = UnitResult(
+                    namespace='code-review',
+                    name='mercurial',
+                    result=UnitResultState.Broken,
+                    details='WARNING: An error occured in the code review bot when applying your patch.\n\n```{}```'.format(e),
+                    format='remarkup',
+                    duration=time.time() - start,
+                )
+
+            if failure is not None:
                 # Remove uncommited changes
                 self.repo.revert(self.repo_dir.encode('utf-8'), all=True)
```

Comment:
Unrelated to this patch but, is this needed? We don't use the repo anymore after this, and new tasks will clean the repo after cloning.

Reply:
We are in pulselistener, sorry, so replace "new tasks will clean the repo after cloning" with "we will clean the repo before attempting to push to try for a new diff".

Reply:
That's one more security i guess. I need to check the behaviour by removing this & crashing patches...

Reply:
Let's file an issue then, we can consider this small improvement in the future.

Reply:
https://github.com/mozilla/release-services/issues/2074